File: api/BeSureApi/Models/ContractHistory.cs
namespace BeSureApi.Models
{
    public class ContractHistory
    {
        public int Id { get; set; }
        public int ContractId { get; set; }
        public string? ContractNumber { get; set; }
        public string CustomerName { get; set; }
        public string AccelLocation { get; set; }
        public string AgreementType { get; set; }
        public string BookingType { get; set; }
        public DateTime BookingDate { get; set; }
        public DateTime BookingValueDate { get; set; }
        public string? QuotationReferenceNumber { get; set; }
        public DateTime? QuotationReferenceDate { get; set; }
        public string? PoNumber { get; set; }
        public DateTime PoDate { get; set; }
        public Decimal ContractValue { get; set; }
        public Decimal AmcValue { get; set; }
        public Decimal FmsValue { get; set; }
        public DateTime StartDate { get; set; }
        public DateTime EndDate { get; set; }
        public Boolean IsMultiSite { get; set; }
        public Boolean IsBackToBackAllowed { get; set; }
        public Boolean IsPmRequired { get; set; }
        public Boolean IsStandByFullUnitRequired { get; set; }
        public Boolean IsStandByImprestStockRequired { get; set; }
        public Boolean IsPreAmcNeeded { get; set; }
        public Boolean IsSez { get; set; }
        public Boolean IsPerformanceGuaranteeRequired { get; set; }
        public string ServiceMode { get; set; }
        public string PaymentMode { get; set; }
        public string PaymentFrequency { get; set; }
        public int SiteCount { get; set; }
        public decimal PerformanceGuaranteeAmount { get; set; }
        public string ServiceWindow { get; set; }
        public string BackToBackScope { get; set; }
        public Boolean NeedPm { get; set; }
        public string PmFrequency { get; set; }
        public string SalesContactPerson { get; set; }
        public DateTime? CallExpiryDate { get; }
        public DateTime? CallStopDate { get; }
        public string? CallStopReason { get; set; }
        public string CreatedBy { get; set; }
        public DateTime CreatedOn { get; set; }
        public DateTime? UpdatedOn { get; set; }
        public string? UpdatedBy { get; set; }
        public DateTime EffectiveFrom{ get; set; }
        public DateTime EffectiveTo { get; set; }
    }

    public class CurrentContract
    {
        public int Id { get; set; }
        public int ContractId { get; set; }
        public string CustomerName { get; set; }
        public string BaseLocation { get; set; }
        public string AgreementType { get; set; }
        public string BookingType { get; set; }
        public DateTime BookingDate { get; set; }
        public DateTime BookingValueDate { get; set; }
        public string? QuotationReferenceNumber { get; set; }
        public string? PoNumber { get; set; }
        public DateTime PoDate { get; set; }
        public Decimal PoValue { get; set; }
        public DateTime StartDate { get; set; }
        public DateTime EndDate { get; set; }
        public Boolean IsMultiSite { get; set; }
        public string ServiceMode { get; set; }
        public string PaymentMode { get; set; }
        public string PaymentFrequency { get; set; }
        public Boolean NeedPm { get; set; }
        public string PmFrequency { get; set; }
        public string SalesContactPerson { get; set; }
        public DateTime CallExpiryDate { get; }
        public DateTime CallStopDate { get; }
        public string? CallStopReason { get; set; }
        public string CreatedBy { get; set; }
        public DateTime CreatedOn { get; set; }
        public DateTime? UpdatedOn { get; set; }
        public string? UpdatedBy { get; set; }
    }
}

File: ui/src/types/contracthistory.ts
import { array, number, Decoder, nullable, object, string, boolean } from 'decoders';

export interface SelectedContractVersion {
    Id: number;
    ContractId: number;
    ContractNumber: string | null;
    CustomerName: string;
    AccelLocation: string;
    AgreementType: string;
    BookingType: string;
    BookingDate: string;
    BookingValueDate: string;
    QuotationReferenceNumber: string | null;
    QuotationReferenceDate: string | null;
    PoNumber: string;
    PoDate: string;
    ContractValue: number | string;
    AmcValue: number | string;
    FmsValue: number | string;
    StartDate: string;
    EndDate: string;
    IsMultiSite: Boolean;
    IsBackToBackAllowed: Boolean;
    IsSez: Boolean,
    SiteCount: number | null;
    ServiceMode: null | string;
    PaymentMode: null | string;
    ServiceWindow: null | string;
    BackToBackScope: null | string;
    PaymentFrequency: null | string;
    PerformanceGuaranteeAmount: number | null;
    IsPmRequired: Boolean | string;
    IsStandByFullUnitRequired: Boolean | string;
    IsStandByImprestStockRequired: Boolean | string;
    IsPreAmcNeeded: Boolean | string;
    IsPerformanceGuaranteeRequired: Boolean,
    PmFrequency: number | string | null;
    SalesContactPerson: null | string;
    CallExpiryDate: string | null;
    CallStopDate: string | null;
    CallStopReason: string | null;
    CreatedOn: string;
    CreatedBy: string;
    UpdatedOn: string | null;
    UpdatedBy: string | null;
    EffectiveFrom: string;
    EffectiveTo: string;
}

export const selectedcontractVersionDecoder: Decoder<SelectedContractVersion> = object({
    Id: number,
    ContractId: number,
    ContractNumber: nullable(string),
    CustomerName: string,
    AccelLocation: string,
    AgreementType: string,
    BookingType: string,
    BookingDate: string,
    BookingValueDate: string,
    QuotationReferenceNumber: nullable(string),
    QuotationReferenceDate: nullable(string),
    PoNumber: string,
    PoDate: string,
    ContractValue: number,
    AmcValue: number,
    FmsValue: number,
    StartDate: string,
    EndDate: string,
    IsMultiSite: boolean,
    IsBackToBackAllowed: boolean,
    SiteCount: nullable(number),
    PerformanceGuaranteeAmount: nullable(number),
    ServiceMode: nullable(string),
    ServiceWindow: nullable(string),
    BackToBackScope: nullable(string),
    PaymentMode: nullable(string),
    PaymentFrequency: nullable(string),
    IsPmRequired: boolean,
    IsStandByFullUnitRequired: boolean,
    IsStandByImprestStockRequired: boolean,
    IsPreAmcNeeded: boolean,
    IsSez: boolean,
    IsPerformanceGuaranteeRequired: boolean,
    PmFrequency: nullable(string),
    SalesContactPerson: nullable(string),
    CallExpiryDate: nullable(string),
    CallStopDate: nullable(string),
    CallStopReason: nullable(string),
    CreatedOn: string,
    CreatedBy: string,
    UpdatedOn: nullable(string),
    UpdatedBy: nullable(string),
    EffectiveFrom: string,
    EffectiveTo: string
});

export interface ContractVersion {
    ContractVersions: SelectedContractVersion[];
    TotalRows: number;
}

export const contractversionDecoder: Decoder<ContractVersion> = object({
    ContractVersions: array(selectedcontractVersionDecoder),
    TotalRows: number
});